### ISSUE TITLE
Simultaneous PDF

### ISSUE THREAD

**Issue by cranmer** [tensorprob/tensorprob#59]
Do you have ability to combine p(x|theta) and p(y|theta) when you have N events for x and M events for y. As in RooSimultaneous?

**Comment by ibab** [comment]
Yes, we're trying to keep the model construction language flexible enough to allow this.
Right now this already works:
```python
import numpy as np
from tensorprob import Model, Normal, Parameter

with Model() as model:
    mu1 = Parameter()
    mu2 = Parameter()
    sigma = Parameter(lower=0)
    X1 = Normal(mu1, sigma)
    X2 = Normal(mu2, sigma)

model.observed(X1, X2)
model.initialize({
    mu1: 2,
    mu2: 2,
    sigma: 2,
})

x1 = np.random.normal(0, 1, size=100)
x2 = np.random.normal(1, 1, size=200)

results = model.fit(x1, x2)
```
Here, I've shared the `sigma` parameter over two datasets of different length.
I'm hoping that something like #23 would make the task of combining different measurements simpler.

**Comment by chrisburr** [comment]
Just to add, you can see [this gist](https://gist.github.com/chrisburr/84774a42b2035cf2fe5d) for a jupyter notebook I made earlier today.

**Comment by cranmer** [comment]
Thanks! Do you have the Poisson terms for extended likelihood fits?

Seems like a good idea to explicitly treat a joint distribution of (x,y)
differently than two extended likelihoods on p(X) and p(y).

**Comment by chrisburr** [comment]
Some kind of wrapper is needed, however, we currently have something that works by defining a `Poisson` distribution where `lambda` is the sum of the two yield parameters. The resulting poisson distributed parameter then becomes an observable in the fit, and needs to be set to the number of data points. You can see an example notebook [here](http://nbviewer.jupyter.org/urls/gist.githubusercontent.com/chrisburr/84774a42b2035cf2fe5d/raw/c41f19ae4cc79a0828190c6ba96c8dd0f1463785/Extended%2520likelihood.ipynb).

**Comment by ibab** [comment]
Yeah, `Mix` is the TensorProb equivalent of `AddPdf`.
This currently allows you to encode the extended likelihood using `Mix` and `Poisson` as building blocks.

It would be nice to treat joint distributions over pairs of numbers differently from joint distributions over different datasets, but I can't think of a nice way of doing it currently.
Maybe we'll come up with something once we spend more time working on multidimensional distributions.

**Comment by cranmer** [comment]
I'm a bit confused. To make the extended likelihood, you would want the equivalent of `ProdPdf(Poisson(n|mean), Shape(x|params))` not a `AddPdf`.  

Would be nice if the Poisson part could automatically get data.size instead of putting it though a separate data entry for N.

I think some think like a dictionary of `{pdf:data}`  is a nice way of organizing the simultaneous pdf.
I'm guessing TensorFlow will be able to figure it out pretty easily, but want that operation to be easy to split across multiple machines/ GPUs etc.

**Comment by ibab** [comment]
By extended likelihood I meant an extended likelihood model with multiple components (with the `AddPdf` being needed to express multiple components). Sorry for the confusion!

In theory it should be trivial to "hide" the `N` parameter from the user by substituting `tf.length(X)` for it.
We just haven't managed to come up with a nice interface for that yet (Would be nice if users could also make use of these substitutions/transformations in their models)

I think the current syntax of
```python
data1 = Pdf1()
data2 = Pdf2()
```
is in a way equivalent to storing things in a `{ pdf: data }` dict and has some advantages.
It can be read as a "generative story" of how the dataset was created, and the user can also easily allocate the calculation to different devices by using functions from tensorflow:
```python
with tf.device("/gpu:0"):
    data1 = Pdf1()
with tf.device("/gpu:1"):
    data2 = Pdf2()
```
